### PR TITLE
Task/cdd 2505 - Multiple Global Banners

### PR DESCRIPTION
**Pull request by luketowell**
# Description

This PR includes the following:

- Removal of the constraint where only 1 Global Banner is available
- Return multiple banners to the front end if they have been marked as active

Fixes #CDD-2505

---

## Type of change

Please select the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Tech debt item (this is focused solely on addressing any relevant technical debt)

---

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests at the right levels to prove my change is effective
- [ ] I have added screenshots or screen grabs where appropriate
- [ ] I have added docstrings in the correct style [(google)](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings)

### CODE REVIEW

File: cms/snippets/serializers/global_banner.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def data(self) -> dict[str, ReturnDict[str, str] | None]:
         active_global_banner_data: ReturnDict[str, str] | None = (
             get_active_global_banner(global_banner_manager=self.global_banner_manager)
         )
-        return {"active_global_banner": active_global_banner_data}
+        return {"active_global_banners": active_global_banner_data}
```
Comment (A-Ashiq) [inline cms/snippets/serializers/global_banner.py:34]: This is introducing a breaking change to an API. We need to bump the global banners endpoint version.
Otherwise soon as you deploy this, the FE will break

File: tests/fakes/managers/cms/global_banner_manager.py [hunk 2]
```diff
@@ -12,12 +12,12 @@ def __init__(self, global_banners: list[GlobalBanner], **kwargs):
...
+    def get_active_banners(self) -> GlobalBanner | None:
         try:
-            return next(
+            return [
                 global_banner
                 for global_banner in self.global_banners
                 if global_banner.is_active is True
-            )
+            ]
         except StopIteration:
```
Comment (A-Ashiq) [inline tests/fakes/managers/cms/global_banner_manager.py:22]: You can remove the catch for the stop iteration error now because it will never get thrown. `StopIteration` gets thrown when an iterator is exhausted i.e. when we call `next()` on something and there's nothing left to give

File: cms/snippets/managers/global_banner.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def has_active_banner(self) -> bool:
         """
         return self.get_queryset().get_active_banners().exists()
 
-    def get_active_banner(self):
+    def get_active_banners(self):
         """Gets the currently active `GlobalBanner`.
```
Comment (A-Ashiq) [inline cms/snippets/managers/global_banner.py:37]:
```suggestion
        """Gets the currently active `GlobalBanner` objects.
```

File: cms/snippets/managers/global_banner.py [hunk 4]
```diff
@@ -42,4 +42,4 @@ def get_active_banner(self):
             then None is returned.
```
Comment (A-Ashiq) [inline cms/snippets/managers/global_banner.py:42]: Can you amend this docstring so it doesn't reference the empty case as None

File: tests/fakes/managers/cms/global_banner_manager.py [hunk 5]
```diff
@@ -12,12 +12,12 @@ def __init__(self, global_banners: list[GlobalBanner], **kwargs):
         self.global_banners = global_banners
         super().__init__(**kwargs)
 
-    def get_active_banner(self) -> GlobalBanner | None:
+    def get_active_banners(self) -> GlobalBanner | None:
```
Comment (A-Ashiq) [inline tests/fakes/managers/cms/global_banner_manager.py:15]:
```suggestion
    def get_active_banners(self) -> list[GlobalBanner]:
```

File: tests/unit/cms/snippets/serializers/test_global_banner.py [hunk 6]
```diff
@@ -78,11 +78,71 @@ def test_serialized_data_for_active_global_banner(self):
...
+                {
+                    "title": active_global_banner.title,
+                    "body": active_global_banner.body,
+                    "banner_type": active_global_banner.banner_type,
+                }
+            ]
+        }
+        assert serializer.data == expected_data
+
+    def test_serialized_data_for_multiple_active_global_banner(self):
```
Comment (A-Ashiq) [inline tests/unit/cms/snippets/serializers/test_global_banner.py:91]: This test is nice 👌

File: cms/snippets/managers/global_banner.py [hunk 7]
```diff
@@ -23,7 +23,7 @@ class GlobalBannerManager(models.Manager):
     def get_queryset(self) -> GlobalBannerQuerySet:
         return GlobalBannerQuerySet(model=self.model, using=self.db)
 
-    def has_active_banner(self) -> bool:
+    def has_active_banners(self) -> bool:
```
Comment (A-Ashiq) [inline cms/snippets/managers/global_banner.py:26]: At this point. Do we have a need for this? Presumably we now don't need to check for if we have any active banners. We can always just plough ahead and make new active banners regardless?

Comment (A-Ashiq) [reply cms/snippets/managers/global_banner.py:26]: In fact is there a need for `get_active_banners()` either?

File: cms/snippets/serializers/global_banner.py (hunk 1 above)
Comment (luketowell) [reply cms/snippets/serializers/global_banner.py:34]: Yea, thats true. I wasn't planning on merging until I had updated the front end. Is endpoint versioning something that has been done elsewhere in the code base? If so can you point me in the direction of an example so I can follow the same standard?

Comment (A-Ashiq) [reply cms/snippets/serializers/global_banner.py:34]: We tend to remove old endpoints from the private API and just keep the ones in use. Since the FE is the only consumer of the private API, it doesn't make sense to keep them around. The main reason we do versioning in the private API is so we have backwards compatilibity for that transition period in which the FE is still pointing to an old version when a new one is available in the BE.
Let's jump on a call and I can talk you through how to do this

File: cms/snippets/serializers/global_banner.py [hunk 8]
```diff
@@ -48,6 +63,26 @@ def get_active_global_banner(
         If no global banner is active, then None is returned
 
     """
-    global_banner = global_banner_manager.get_active_banner()
-    serializer = GlobalBannerResponseSerializer(instance=global_banner)
+    global_banners = global_banner_manager.get_active_banners().first()
```
Comment (A-Ashiq) [inline cms/snippets/serializers/global_banner.py:66]: The reason we use model managers is so that we can push this kind of stuff over into the model manager abstraction. So we don't leak that implementation detail into the caller.
I don't mind if you leave this one as you have it here, cause we'll be deleting this soon anyway, but just something to bear in mind

File: cms/snippets/managers/global_banner.py (hunk 7 above)
Comment (A-Ashiq) [inline cms/snippets/managers/global_banner.py:26]: At this point, do we need this?

File: tests/fakes/managers/cms/global_banner_manager.py (hunk 5 above)
Comment (A-Ashiq) [inline tests/fakes/managers/cms/global_banner_manager.py:15]: For `get_active_banners` we are probably gonna return `-> list[GlobalBanner]`

File: tests/fakes/managers/cms/global_banner_manager.py (hunk 2 above)
Comment (A-Ashiq) [inline tests/fakes/managers/cms/global_banner_manager.py:22]: We would no longer expect this `StopIteration` error to ever be thrown. In the case of no global banners, we would just return an empty list. 
The `StopIteration` error gets thrown when we exhaust an iterable, which we would have done with the `next()` keyword

File: tests/integration/cms/snippets/views/test_global_banner.py [hunk 9]
```diff
@@ -11,13 +11,13 @@
 class TestGlobalBannerView:
     @property
     def path(self) -> str:
-        return "/api/global-banners/v1"
+        return "/api/global-banners/v2"
```
Comment (A-Ashiq) [inline tests/integration/cms/snippets/views/test_global_banner.py:14]: Same comment as above. We need to maintain the test coverage we had before on the deprecated endpoint for the time being.

File: cms/snippets/serializers/global_banner.py [hunk 10]
```diff
@@ -48,6 +63,26 @@ def get_active_global_banner(
...
+    serializer = GlobalBannerResponseSerializer(instance=global_banners)
+    return serializer.data
+
+
+def get_active_global_banners(
+    global_banner_manager: Manager,
+) -> ReturnDict[str, str] | None:
```
Comment (A-Ashiq) [inline cms/snippets/serializers/global_banner.py:73]: Do we ever return None?

File: cms/snippets/serializers/global_banner.py [hunk 11]
```diff
@@ -48,6 +63,26 @@ def get_active_global_banner(
...
+    """Gets the currently active global banner information
```
Comment (A-Ashiq) [inline cms/snippets/serializers/global_banner.py:74]:
```suggestion
    """Gets information about any currently active global banners
```

File: cms/snippets/serializers/global_banner.py [hunk 12]
```diff
@@ -48,6 +63,26 @@ def get_active_global_banner(
...
+
+    Args:
+        `global_banner_manager`: The `GlobalBanner`
+            model manager used to query for records
+
+    Returns:
+        Dict representation the of the active global banner.
```
Comment (A-Ashiq) [inline cms/snippets/serializers/global_banner.py:81]:
```suggestion
        Dict representation of any active global banners.
```

File: cms/snippets/serializers/global_banner.py [hunk 13]
```diff
@@ -48,6 +63,26 @@ def get_active_global_banner(
...
+        If no global banner is active, then None is returned
```
Comment (A-Ashiq) [inline cms/snippets/serializers/global_banner.py:82]: Should this be an empty list not None?

File: tests/integration/cms/snippets/managers/test_global_banner.py [hunk 14]
```diff
@@ -18,7 +18,7 @@ def test_has_active_banner(self):
         GlobalBannerFactory.create(is_active=True)
 
         # When
-        has_active_banner: bool = GlobalBanner.objects.has_active_banner()
+        has_active_banners: bool = GlobalBanner.objects.has_active_banners()
```
Comment (A-Ashiq) [inline tests/integration/cms/snippets/managers/test_global_banner.py:21]: Same comment as before. I'm not sure we need to use `has_active_banners()` anymore. That was only ever used as a validation check to enfore a single active banner at any given time, but that's no longer a restriction we should be enforcing

File: tests/unit/cms/snippets/models/test_global_banner.py [hunk 15]
```diff
@@ -108,99 +107,3 @@ def test_inactive_global_banner_produces_correct_dunder_str(self):
         current_active_banner_label = "Active information-level global banner"
         assert current_active_banner_label not in string_representation
         assert "Inactive" in string_representation
-
-    @mock.patch.object(GlobalBannerManager, "has_active_banner")
```
Comment (A-Ashiq) [inline tests/unit/cms/snippets/models/test_global_banner.py:112]: These tests are fine to be removed 👍

File: tests/unit/cms/snippets/serializers/test_global_banner.py [hunk 16]
```diff
@@ -119,5 +179,5 @@ def test_serialized_data_for_no_active_global_banner(self):
         )
 
         # Then
-        expected_data = {"active_global_banner": None}
+        expected_data = {"active_global_banners": []}
```
Comment (A-Ashiq) [inline tests/unit/cms/snippets/serializers/test_global_banner.py:182]: You need to update the test docstring to reference the empty list as the expected return value, no longer None

File: cms/snippets/serializers/global_banner.py (hunk 8 above)
Comment (A-Ashiq) [inline cms/snippets/serializers/global_banner.py:66]: **Nitpick**: the plurality of this variable leads the reader to think they're getting back multiple objects. But in this case we're only getting either 1 or None

Comment (luketowell) [reply cms/snippets/serializers/global_banner.py:66]: The reason I pulled it out of the manager is so that I can change the manager to return the array and then we can temporarily manipulate the array in the serializer until we get rid of him.

File: cms/snippets/managers/global_banner.py (hunk 7 above)
Comment (luketowell) [reply cms/snippets/managers/global_banner.py]: No, I have now removed it and updated the tests.

File: tests/fakes/managers/cms/global_banner_manager.py (hunk 2 above)
Comment (luketowell) [reply tests/fakes/managers/cms/global_banner_manager.py]: 👍

File: cms/snippets/serializers/global_banner.py (hunk 10 above)
Comment (luketowell) [reply cms/snippets/serializers/global_banner.py]: Good catch, this has been resolved

File: cms/snippets/serializers/global_banner.py (hunk 8 above)
Comment (luketowell) [reply cms/snippets/serializers/global_banner.py]: Updated now due to adding the initial endpoint back in to the application.

File: tests/unit/cms/snippets/serializers/test_global_banner.py (hunk 16 above)
Comment (luketowell) [reply tests/unit/cms/snippets/serializers/test_global_banner.py:182]: Updated 👍

File: tests/integration/cms/snippets/managers/test_global_banner.py (hunk 14 above)
Comment (luketowell) [reply tests/integration/cms/snippets/managers/test_global_banner.py]: Updated

File: cms/snippets/serializers/global_banner.py (hunk 13 above)
Comment (luketowell) [reply cms/snippets/serializers/global_banner.py]: Updated

File: cms/snippets/serializers/global_banner.py (hunk 12 above)
Comment (luketowell) [reply cms/snippets/serializers/global_banner.py]: Updated

File: cms/snippets/serializers/global_banner.py (hunk 11 above)
Comment (luketowell) [reply cms/snippets/serializers/global_banner.py]: Updated

File: cms/snippets/views/global_banner.py [hunk 17]
```diff
@@ -37,3 +38,30 @@ def get(cls, request, *args, **kwargs) -> Response:
         )
```
Comment (A-Ashiq) [inline cms/snippets/views/global_banner.py:38]: You will need to update the docstring in the v2 view banner
Screenshot 2025-03-11 at 16 14 26

File: tests/integration/cms/snippets/managers/test_global_banner.py [hunk 18]
```diff
@@ -6,7 +6,7 @@
 
 class TestGlobalBannerManager:
     @pytest.mark.django_db
-    def test_has_active_banner(self):
+    def test_get_active_banners(self):
```
Comment (A-Ashiq) [inline tests/integration/cms/snippets/managers/test_global_banner.py:9]: This test needs a re-write. The docstring still talks about returning True from when we checked banners existed.
But the assertion also is now checking its not None. But it will never be None. In the false condition it will be empty list

**Comment by A-Ashiq** [comment]
@luketowell you will also need to update the help text which gets rendered on the CMS for the `is_active` field on the GlobalBanner snippet model:
cms/dynamic_content/help_texts.py

File: cms/dynamic_content/help_texts.py [hunk 19]
```diff
@@ -297,9 +297,8 @@
 
 GLOBAL_BANNER_IS_ACTIVE: str = """
 Whether to activate this banner across every page in the dashboard. 
-Note that only 1 global banner can be active at a time.
-To switch from 1 active banner straight to another, 
-you must deactivate the 1st banner and save it before activating and saving the 2nd banner.
+Multiple banners can be active across the dashboard.
+Note: Warning banners are prioritised across the dashboard and will be displayed above information banners.
```
Comment (A-Ashiq) [inline cms/dynamic_content/help_texts.py:301]: Out of curiosity how would we prioritise banners within each subgroup.
E.g. if we have 2 warning banners and another information banner all active, we would have:
```
Warning
Warning 
Information
```
But how do we decide the order of the warnings? Is this going to be handled by the frontend?

Comment (luketowell) [reply cms/dynamic_content/help_texts.py:301]: This is a really good question, the ordering of warning and information groupings is straightforward however if we want to order those groupings by created date or something similar then we will need to add a field to the model so that we can subsequently order by that date time field.
